### ISSUE TITLE
[macos mojave 10.14] compiling c++ files error

### ISSUE THREAD

**Issue by kkshmz** [panda3d/panda3d#657]
Hi I am running an error that I can't compile panda3d c++ files.
I am currently trying to compile this [repo](https://github.com/lettier/3d-game-shaders-for-beginners) but when compiling I get an error like.
```
Apple LLVM version 10.0.1 (clang-1001.0.46.3)
Target: x86_64-apple-darwin18.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.14.0 -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -o 3d-game-shaders-for-beginners -L/Developer/Panda3D/lib 3d-game-shaders-for-beginners.o -lp3framework -lpanda -lpandafx -lpandaexpress -lp3dtoolconfig -lp3dtool -lp3pystub -lp3direct -lpthread -L/usr/local/lib -lc++ -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.1/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture x86_64:
  "AmbientLight::AmbientLight(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      generateLights(NodePath, bool) in 3d-game-shaders-for-beginners.o
  "InternalName::append(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      PointerTo<InternalName> InternalName::make<16>(char const (&) [16]) in 3d-game-shaders-for-beginners.o
      PointerTo<InternalName> InternalName::make<14>(char const (&) [14]) in 3d-game-shaders-for-beginners.o
      PointerTo<InternalName> InternalName::make<8>(char const (&) [8]) in 3d-game-shaders-for-beginners.o
      PointerTo<InternalName> InternalName::make<6>(char const (&) [6]) in 3d-game-shaders-for-beginners.o
      PointerTo<InternalName> InternalName::make<15>(char const (&) [15]) in 3d-game-shaders-for-beginners.o
      PointerTo<InternalName> InternalName::make<12>(char const (&) [12]) in 3d-game-shaders-for-beginners.o
      PointerTo<InternalName> InternalName::make<23>(char const (&) [23]) in 3d-game-shaders-for-beginners.o
      ...
  "InternalName::InternalName(InternalName*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      InternalName::get_root() in 3d-game-shaders-for-beginners.o
  "TypeRegistry::register_type(TypeHandle&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      PointerToBase<ReferenceCount>::update_type(ReferenceCount*) in 3d-game-shaders-for-beginners.o
      NodeReferenceCount::init_type() in 3d-game-shaders-for-beginners.o
      AsyncTaskManager::init_type() in 3d-game-shaders-for-beginners.o
      TypedReferenceCount::init_type() in 3d-game-shaders-for-beginners.o
      WindowFramework::init_type() in 3d-game-shaders-for-beginners.o
      TypedWritableReferenceCount::init_type() in 3d-game-shaders-for-beginners.o
      TypedWritable::init_type() in 3d-game-shaders-for-beginners.o
      ...
  "TypeRegistry::record_alternate_name(TypeHandle, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      TypedWritableReferenceCount::init_type() in 3d-game-shaders-for-beginners.o
      TypedWritable::init_type() in 3d-game-shaders-for-beginners.o
  "TypeRegistry::register_dynamic_type(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      CopyOnWriteObj1<ov_multiset<PandaNode::DownConnection, std::__1::less<PandaNode::DownConnection>, pvector<PandaNode::DownConnection> >, TypeHandle>::init_type() in 3d-game-shaders-for-beginners.o
      CopyOnWriteObj1<ov_set<PandaNode::UpConnection, std::__1::less<PandaNode::UpConnection>, pvector<PandaNode::UpConnection> >, TypeHandle>::init_type() in 3d-game-shaders-for-beginners.o
  "ButtonRegistry::find_button(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      isButtonDown(WindowFramework*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) in 3d-game-shaders-for-beginners.o
  "GraphicsEngine::make_output(GraphicsPipe*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, FrameBufferProperties const&, WindowProperties const&, int, GraphicsStateGuardian*, GraphicsOutput*)", referenced from:
      generateHighPrecisionFramebufferTextureForScene(PointerTo<WindowFramework>, PointerTo<GraphicsOutput>, PointerTo<GraphicsEngine>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) in 3d-game-shaders-for-beginners.o
  "GraphicsOutput::make_texture_buffer(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, Texture*, bool, FrameBufferProperties*)", referenced from:
      generateFramebufferTextureForImage(PointerTo<GraphicsOutput>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LVecBase4f) in 3d-game-shaders-for-beginners.o
      generateFramebufferTextureForScene(PointerTo<WindowFramework>, PointerTo<GraphicsOutput>, PointerTo<GraphicsEngine>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LVecBase4f) in 3d-game-shaders-for-beginners.o
  "DirectionalLight::DirectionalLight(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      generateLights(NodePath, bool) in 3d-game-shaders-for-beginners.o
  "GenericAsyncTask::GenericAsyncTask(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, AsyncTask::DoneStatus (*)(GenericAsyncTask*, void*), void*)", referenced from:
      _main in 3d-game-shaders-for-beginners.o
  "Fog::Fog(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      _main in 3d-game-shaders-for-beginners.o
  "Camera::Camera(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Lens*)", referenced from:
      generateFramebufferTextureForImage(PointerTo<GraphicsOutput>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LVecBase4f) in 3d-game-shaders-for-beginners.o
  "PandaNode::PandaNode(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      _main in 3d-game-shaders-for-beginners.o
      NodePath::NodePath(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Thread*) in 3d-game-shaders-for-beginners.o
  "Spotlight::Spotlight(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      generateWindowLight(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, NodePath, LVecBase3f, bool) in 3d-game-shaders-for-beginners.o
  "NodePath::find(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const", referenced from:
      _main in 3d-game-shaders-for-beginners.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```
Trying again installing from source, I get an error about openAL, but it seems resolved in this PR. Is it just a matter of upgrading my enviornment?
https://github.com/panda3d/panda3d/pull/596

My enviornment is 
macos mojave 10.14.3
xcode 10.2 
and installed through the dmg file provided in the githubrepo

**Comment by rdb** [comment]
Maybe you need to add the `-std=gnu++11` flag to enable compiling with C++11 features enabled.

**Comment by kkshmz** [comment]
Because I have installed the Panda3d through the DMG, I clicked on enabling the c++ headers.
But are you suggesting that I should install through source with some arguement for compiling with c++11 features?

When I compile the program for the repo, I do explicit -std=gnu+11 as so
```
g++ \
  -c main.cxx \
  -o 3d-game-shaders-for-beginners.o \
  -std=gnu++11 \
  -O2 \
  -I/usr/include/python2.7/ \
  -I/usr/include/panda3d/
```

**Comment by rdb** [comment]
I'm going to close this as this is a support request and not actionable as an issue.  Are you still struggling with this?  I would certainly recommend that you use clang++ instead of g++, since g++ on macOS is very outdated, double-check that the Panda3D libraries are properly specified on the linker command-line, and double-check that you are using the same stdlib as your build of Panda3D.

**Comment by rdb** [comment]
I wanted to mention for the benefit of anyone coming upon this issue through Google, that this issue is now properly addressed in 1.10.5:
https://www.panda3d.org/blog/sdk-1-10-5-release/
